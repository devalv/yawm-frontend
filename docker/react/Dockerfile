FROM node:17-slim

ENV APP_HOME /home/app
ENV CHOKIDAR_USEPOLLING true
ENV NODE_OPTIONS "--openssl-legacy-provider"
ENV REACT_APP_API_ORIGIN=https://domain.site
ENV REACT_APP_API_URL=https://domain.site/api
ENV REACT_APP_API_V1_URL=https://domain.site/api/v1
ENV REACT_APP_API_V2_URL=https://domain.site/api/v2
ENV REACT_APP_HEALTH_URL=https://domain.site/api/health
ENV REACT_APP_STATE=nonrandom-docker-state
ENV REACT_APP_SENTRY_DSN="<secret>"

RUN addgroup --system app && adduser --system --group app

RUN mkdir -p $APP_HOME

WORKDIR $APP_HOME

COPY ./frontend/ ./
RUN npm install

RUN chown -R app:app $APP_HOME

USER app
RUN npm run build
