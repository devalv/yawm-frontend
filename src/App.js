import React, { Component } from 'react';


class App extends Component {
// TODO: @devalv read env
  state = {
    producerLoginRedirectEndpoint: 'http://localhost:8000/v1/react_login/',
    producerLoginEndpoint: 'http://localhost:8000/v1/react_swap_token/',
    producerLogoutEndpoint: 'http://localhost:8000/v1/logout/',
    producerLoginCheckEndpoint: 'http://localhost:8000/v1/user/info/',
    userLoggedIn: false,
    userName: null,
  };

  componentDidMount() {
    this.authenticate()
  };

  setCookie = (cname, cvalue, exdays) => {
    console.log('setting cookie');
    var d = new Date();
    d.setTime(d.getTime() + (exdays*24*60*60*1000));
    var expires = "expires="+ d.toUTCString();
    document.cookie = cname + "=" + cvalue + ";" + expires + ";path=/";
    console.log('cookie was set');
  };

  getCookie = (cname) => {
    console.log('getting cookie');
    var name = cname + "=";
    var decodedCookie = decodeURIComponent(document.cookie);
    var ca = decodedCookie.split(';');
    for(var i = 0; i <ca.length; i++) {
      var c = ca[i];
      while (c.charAt(0) === ' ') {
        c = c.substring(1);
      }
      if (c.indexOf(name) === 0) {
        return c.substring(name.length, c.length);
      }
    }
    return "";
  }

  authenticate = () => {
    console.log('authenticate called');
    let accessToken = (window.location.search.match(/authToken=([^&]+)/) || [])[1];
    
    if (!accessToken) {
        console.log('there is no access token on location');
        accessToken = this.getCookie('access_token');
        console.log('token from cookie:', accessToken);
    };

    if (accessToken) {
      // Try to get an access token from the server      
        console.log('received token from page');
        this.checkUserSessionStatus(accessToken);
        this.setCookie('access_token', accessToken);
        window.history.pushState('object', document.title, "/");

    } else {
      // Check user is logged in
        console.log('Can`t find token anywhere.');
    }
  }

  getAccessToken = (accessToken) => {
    const request = {
      method: 'GET',
      headers: {
        "Authorization": "Bearer " + accessToken
      },
      credentials: 'include'
    }

    fetch(this.state.producerLoginEndpoint, request)
    .then(response => {
      // Check user is logged in
      this.checkUserSessionStatus()
    })
    .then(data => {})
    .catch(err => {})
  }

  checkUserSessionStatus = (accessToken) => {
    console.log('access token:', accessToken);
    const request = {
        method: 'GET',
        headers: {
        "Authorization": "Bearer " + accessToken
        },
        credentials: 'include'
    };

    fetch(this.state.producerLoginCheckEndpoint, request)
    .then(response => {
        if(!response.ok) throw new Error(response.status);
        else return response.json();
      })
    .then(data => {
      this.setState({
        userLoggedIn: true,
        userName: data['username'],
      });
    })
    .catch(err => {})
  }

  logout = () => {
    const request = {
      method: 'GET',
      headers: {
        "Authorization": "Bearer " + this.getCookie('access_token')
      },
      credentials: 'include'
    }

    fetch(this.state.producerLogoutEndpoint, request)
    .then(response => {
        if(!response.ok) throw new Error(response.status);
        else {this.setCookie('access_token', null); window.location.reload();}
      })
    .catch(err => {})
  }

  render() {
    return (
      <section id="page-container">
        {this.state.userLoggedIn ?
          <div>
            <div>
              {this.state.userName} you are now logged in!
            </div>
            <div>
              <button onClick={this.logout}>Logout</button>
            </div>
          </div> :
          <Login producerLoginRedirectEndpoint={this.state.producerLoginRedirectEndpoint}/>
        }
      </section>
    );
  }
}


function Login(props) {
  const googleLogin = () => {
      // TODO: @devalv read state from env
    var auth_state = "react-random-str"
    var login_url = props.producerLoginRedirectEndpoint + "?state=" + auth_state
    window.location.href = login_url
  }

  return (
    <section>
      <div>
        <button onClick={googleLogin}>Login with Google</button>
      </div>
    </section>
  );
}

export default App;