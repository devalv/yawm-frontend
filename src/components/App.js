import * as React from "react";
import { BrowserRouter, Route, Routes } from "react-router-dom";
import axios from 'axios';
import Cookies from 'js-cookie';
import { ErrorContext, errorState, AuthContext, authState } from './GlobalContext';
import { simpleLogout } from './auth/Logout';
import ErrorModal from './ErrorModal';
import ButtonAppBar from './Navbar';
import NestedWishlistsGrid from "./wishlist/list/WishlistsGridList";
// import DetailWishlistCard from "./wishlist/detailWishlistCard";


function App() {
  // Errors
  const [ErrorState, setErrorState] = React.useState(errorState.noErrors);
  const ErrorsContextProviderValue = React.useMemo(() => ({ErrorState, setErrorState}), [ErrorState]);

  // Auth
  const accessTokenCookie = Cookies.get("access_token");
  let authStateInitial = authState.anonymousUser;
  if (accessTokenCookie) {
    authState.loggedInUser.accessToken = accessTokenCookie;
    authStateInitial = authState.loggedInUser;
  }
  const [AuthState, setAuthState] = React.useState(authStateInitial);
  const AuthContextProviderValue = React.useMemo(() => ({AuthState, setAuthState}), [AuthState]);

  // Interceptors
  axios.interceptors.response.use((response) => response, (error) => {
    // Any status codes that falls outside the range of 2xx cause this function to trigger
    if (error.response.status === 401) {
      simpleLogout();
    } else if (error.response.status !== 400) {
      setErrorState(errorState.anyError);
    }
    return Promise.reject(error);
  });

  // Effects
  React.useEffect(() => {
    const getUserInfo = async () => {
      const { REACT_APP_API_V2_URL } = process.env;
      const userInfoEndpoint = `${REACT_APP_API_V2_URL}/users/info`;
      await axios.get(userInfoEndpoint).then((response) => {
        authState.loggedInUser.userId = response.data.id;
        authState.loggedInUser.username = response.data.username;
        setAuthState(authState.loggedInUser);
        // TODO: @devalv все говорит о том, что придется вкорячить CSRF и слать в целом не через заголовки, а через куки + использовать httpOnly: true?
        Cookies.set("access_token", authState.loggedInUser.accessToken, { expires: 1/48, secure: true , sameSite: 'strict' });
      }).catch(() => {
        simpleLogout();
      });
    };
    if ((AuthState.authenticated === true) && (AuthState.accessToken)) {
      // Навешиваем данные о токене во всех возможные запросы
      axios.defaults.headers.common.Authorization = `Bearer ${AuthState.accessToken}`;
      // Прочитать информацию о пользователе
      getUserInfo();
    }
  }, [AuthState]);

  return (
    <ErrorContext.Provider value={ErrorsContextProviderValue}>
      <AuthContext.Provider value={AuthContextProviderValue}>
        <ButtonAppBar/>
        <BrowserRouter>
          <Routes>
            <Route path="/" element={<NestedWishlistsGrid />}/>
            {/* <Route */}
            {/*  path="/wishlist/:id" */}
            {/*  element={<DetailWishlistCard userId={AuthState.user_id} />} */}
            {/* /> */}
          </Routes>
        </BrowserRouter>
        <ErrorModal />
      </AuthContext.Provider>
     </ErrorContext.Provider>
  );
}

export default App;
