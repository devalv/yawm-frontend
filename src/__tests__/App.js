import * as React from 'react';
import renderer from 'react-test-renderer';

import App from '../components/App';

it('check that index page can be rendered', () => {
  const tree = renderer.create(<App />).toJSON();
  expect(tree).toMatchSnapshot();
});
