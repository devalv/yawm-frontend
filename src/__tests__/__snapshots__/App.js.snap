// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check that index page can be rendered 1`] = `
Array [
  <div
    className="MuiBox-root css-15lzmko"
  >
    <header
      className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
    >
      <div
        className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
      >
        <button
          aria-label="menu"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeLarge css-ancrnh-MuiButtonBase-root-MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="GitHubIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 1.27a11 11 0 00-3.48 21.46c.55.09.73-.28.73-.55v-1.84c-3.03.64-3.67-1.46-3.67-1.46-.55-1.29-1.28-1.65-1.28-1.65-.92-.65.1-.65.1-.65 1.1 0 1.73 1.1 1.73 1.1.92 1.65 2.57 1.2 3.21.92a2 2 0 01.64-1.47c-2.47-.27-5.04-1.19-5.04-5.5 0-1.1.46-2.1 1.2-2.84a3.76 3.76 0 010-2.93s.91-.28 3.11 1.1c1.8-.49 3.7-.49 5.5 0 2.1-1.38 3.02-1.1 3.02-1.1a3.76 3.76 0 010 2.93c.83.74 1.2 1.74 1.2 2.94 0 4.21-2.57 5.13-5.04 5.4.45.37.82.92.82 2.02v3.03c0 .27.1.64.73.55A11 11 0 0012 1.27"
            />
          </svg>
        </button>
        <a
          aria-label="menu"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeLarge css-ancrnh-MuiButtonBase-root-MuiIconButton-root"
          href="/"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
        >
          <img
            alt=""
            height="32"
            src="/logo64.png"
            width="32"
          />
        </a>
        <div
          className="MuiTypography-root MuiTypography-h6 css-1juivf6-MuiTypography-root"
        >
          Мастер списков
        </div>
        <button
          className="MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit MuiButtonBase-root  css-1y942vo-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          Войти
        </button>
      </div>
    </header>
  </div>,
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
  >
    <div
      className="MuiBox-root css-i9gxme"
    >
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 css-a5rdam-MuiGrid-root"
      >
        <div
          className="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-spacing-xs-2 MuiGrid-spacing-md-4 MuiGrid-spacing-lg-6 css-1yzxgpl-MuiGrid-root"
        />
      </div>
    </div>
    <br />
    <div
      className="css-1p5q5e5-MuiStack-root"
    >
      <nav
        aria-label="pagination navigation"
        className="MuiPagination-root MuiPagination-text css-1oj2twp-MuiPagination-root"
      >
        <ul
          className="MuiPagination-ul css-wjh20t-MuiPagination-ul"
        >
          <li>
            <button
              aria-current="true"
              aria-label="page 1"
              className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-textPrimary Mui-selected MuiPaginationItem-page css-1to7aaw-MuiButtonBase-root-MuiPaginationItem-root"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              1
            </button>
          </li>
        </ul>
      </nav>
    </div>
  </div>,
  <div />,
]
`;
